File: Components/Strata.Foundation.Core/main/typescript/Value/PersonNameBuilder.ts
import {PersonName} from "./PersonName";

export
class PersonNameBuilder
{
    private title?: string;
    private firstName: string;
    private middleName?: string;
    private lastName: string;
    private suffix?: string;

    constructor()
    {
        this.title = undefined;
        this.firstName = null;
        this.middleName = undefined;
        this.lastName = null;
        this.suffix = undefined;
    }

    setTitle(title: string): PersonNameBuilder
    {
        this.title = title;
        return this;
    }

    setFirstName(firstName: string): PersonNameBuilder
    {
        this.firstName = firstName;
        return this;
    }

    setMiddleName(middleName: string): PersonNameBuilder
    {
        this.middleName = middleName;
        return this;
    }

    setLastName(lastName: string): PersonNameBuilder
    {
        this.lastName = lastName;
        return this;
    }

    setSuffix(suffix: string): PersonNameBuilder
    {
        this.suffix = suffix;
        return this;
    }

    clear(): PersonNameBuilder
    {
        this.title = undefined;
        this.firstName = null;
        this.middleName = undefined;
        this.lastName = null;
        this.suffix = undefined;
        return this;
    }

    getTitle(): string|undefined
    {
        return this.title;
    }

    getFirstName(): string
    {
        return this.firstName;
    }

    getMiddleName(): string|undefined
    {
        return this.middleName;
    }

    getLastName(): string
    {
        return this.lastName;
    }

    getSuffix(): string|undefined
    {
        return this.suffix;
    }

    hasTitle(): boolean
    {
        return this.title != undefined;
    }

    hasFirstName(): boolean
    {
        return this.firstName != null;
    }

    hasMiddleName(): boolean
    {
        return this.middleName != undefined;
    }

    hasLastName(): boolean
    {
        return this.lastName != null;
    }

    hasSuffix(): boolean
    {
        return this.suffix != undefined;
    }

    build(): PersonName
    {
        return {
            title: this.title,
            firstName: this.firstName,
            middleName: this.middleName,
            lastName: this.lastName,
            suffix: this.suffix
        };
    }
}

File: Components/Strata.Foundation.Core/main/java/strata/foundation/core/value/PersonNameBuilder.java
//////////////////////////////////////////////////////////////////////////////
// PersonNameBuilder.java
//////////////////////////////////////////////////////////////////////////////

package strata.foundation.core.value;

import java.util.Objects;
import java.util.Optional;

/**
 * <p>
 * A mutable <a href="https://en.wikipedia.org/wiki/Builder_pattern">builder</a>
 * for incrementally assembling a {@link PersonName}. Each {@code setXxx}
 * method returns {@code this}, allowing calls to be chained fluently, and
 * {@code hasXxx}/{@code getXxx} accessors let callers inspect the fields that
 * have been set so far before calling {@link #build()} to produce the
 * resulting immutable {@link PersonName}. A first name and last name must be
 * set before {@link #build()} is called.
 * </p><br/>
 * <b>Examples</b><br/>
 * <pre>
 * PersonName name =
 *     new PersonNameBuilder()
 *         .setTitle("Dr.")
 *         .setFirstName("Jane")
 *         .setMiddleName("Q.")
 *         .setLastName("Doe")
 *         .setSuffix("Jr.")
 *         .build();
 * </pre>
 */
public
class PersonNameBuilder
{
    private String itsTitle;
    private String itsFirstName;
    private String itsMiddleName;
    private String itsLastName;
    private String itsSuffix;

    public
    PersonNameBuilder()
    {
        itsTitle = null;
        itsFirstName = null;
        itsMiddleName = null;
        itsLastName = null;
        itsSuffix = null;
    }

    public PersonNameBuilder
    setTitle(String title)
    {
        itsTitle = title;
        return this;
    }

    public PersonNameBuilder
    setFirstName(String firstName)
    {
        itsFirstName = firstName;
        return this;
    }

    public PersonNameBuilder
    setMiddleName(String middleName)
    {
        itsMiddleName = middleName;
        return this;
    }

    public PersonNameBuilder
    setLastName(String lastName)
    {
        itsLastName = lastName;
        return this;
    }

    public PersonNameBuilder
    setSuffix(String suffix)
    {
        itsSuffix = suffix;
        return this;
    }

    public PersonNameBuilder
    clear()
    {
        itsTitle = null;
        itsFirstName = null;
        itsMiddleName = null;
        itsLastName = null;
        itsSuffix = null;
        return this;
    }

    public Optional<String>
    getTitle() { return Optional.ofNullable(itsTitle); }

    public String
    getFirstName() { return Objects.requireNonNull(itsFirstName); }

    public Optional<String>
    getMiddleName() { return Optional.ofNullable(itsMiddleName); }

    public String
    getLastName() { return Objects.requireNonNull(itsLastName); }

    public Optional<String>
    getSuffix() { return Optional.ofNullable(itsSuffix); }

    public boolean
    hasTitle() { return Objects.nonNull(itsTitle); }

    public boolean
    hasFirstName() { return Objects.nonNull(itsFirstName); }

    public boolean
    hasMiddleName() { return Objects.nonNull(itsMiddleName); }

    public boolean
    hasLastName() { return Objects.nonNull(itsLastName); }

    public boolean
    hasSuffix() { return Objects.nonNull(itsSuffix); }

    public PersonName
    build()
    {
        return
            new PersonName(
                itsTitle,
                itsFirstName,
                itsMiddleName,
                itsLastName,
                itsSuffix);
    }
}

//////////////////////////////////////////////////////////////////////////////
